[quantization] equal() for QuantizedCPU

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #23987 [quantization] JIT trace testing
* #24298 [JIT] Add the ability to compile exports on traced modules
* **#24211 [quantization] equal() for QuantizedCPU**

Differential Revision: [D16799801](https://our.internmc.facebook.com/intern/diff/D16799801)

Commit:
[quantization] equal() for QuantizedCPU

Commit:
Update on "[quantization] equal() for QuantizedCPU"

Commit:
Update on "[quantization] equal() for QuantizedCPU"

zafartahirov commented on aten/src/ATen/native/quantized/QTensor.cpp:
```diff
@@ -151,5 +151,27 @@ Tensor quantized_clone(const Tensor& self) {
   return dst;
 }
 
+bool quantized_equal(const Tensor& self, const Tensor& other) {
```
Do we define equality as the tensors being absolutely equal, or numerically equal? If the latter is true, we need to check if dequantized values are equal, no matter the scale/zero_point.

jamesr66a replied:
@zafartahirov we need both. This quick implementation is used for situations where we are checking tensor identity for memoization, which is currently implemented by calling equal(). It is an error to mark tensors that are numerically equivalent in the reals but have differing quantization parameters in this scenario.

zafartahirov commented on aten/src/ATen/native/quantized/QTensor.cpp:
```diff
@@ -151,5 +151,27 @@ Tensor quantized_clone(const Tensor& self) {
   return dst;
 }
 
+bool quantized_equal(const Tensor& self, const Tensor& other) {
```
We need a check for the scheme as well:

```python
if (self.q_scheme() != other.q_scheme) return false;
```

jamesr66a replied:
I think I'm actually going to define comparisons on Quantizer and delegate to those. That way we can have specific logic for each scheme to check equivalence (i.e. comparing scale vs scales for per_channel)

Commit:
Update on "[quantization] equal() for QuantizedCPU"

Commit:
Update on "[quantization] equal() for QuantizedCPU"

Commit:
Update on "[quantization] equal() for QuantizedCPU"

Commit:
Update on "[quantization] equal() for QuantizedCPU"

z-a-f reviewed (approved):
LGTM!

Commit:
Update on "[quantization] equal() for QuantizedCPU"

Commit:
Update on "[quantization] equal() for QuantizedCPU"

jerryzh168 commented on aten/src/ATen/native/quantized/QTensor.cpp:
```diff
@@ -151,5 +151,35 @@ Tensor quantized_clone(const Tensor& self) {
+
+  void* self_data = self_contig.data_ptr();
+  void* other_data = other_contig.data_ptr();
+  return 0 == memcmp(self_data, other_data, self.numel() * self.element_size());
```
maybe just compare int_repr()?

jamesr66a replied:
int_repr looks like it does a memcpy, which is probably unnecessary here

jerryzh168 replied:
OK

zafartahirov replied:
Although no needed for this PR, do you by any chance know how does `memcmp` compare to `TensorIterator` + value comparison? My assumption was that the iterators have multi=threading in their backend. Not sure if the memcmp has the same

Commit:
Update on "[quantization] equal() for QuantizedCPU"


Differential Revision: [D16799801](https://our.internmc.facebook.com/intern/diff/D16799801)

jerryzh168 commented on test/test_quantized.py:
```diff
@@ -341,6 +341,62 @@ def test_cat(self, X, num, axis, relu):
+                                        dtype=torch_type2)
+
+        def equal_ref(X, params, X_scheme, X2, params2, X2_scheme):
+            if X_scheme != X2_scheme:
```
I think you can just call X.qscheme() to get the qscheme?

jamesr66a replied:
These are the float tensors, not the quantized tensors

jerryzh168 commented on aten/src/ATen/native/quantized/QTensor.cpp:
```diff
@@ -151,5 +151,35 @@ Tensor quantized_clone(const Tensor& self) {
   return dst;
 }
 
+bool quantized_equal(const Tensor& self, const Tensor& other) {
```
nit: right now we use naming convention: "xxxx_quant" for quantized impelementations. so should be "equal_quant"

jamesr66a replied:
That's not the case for `quantized_{conv,relu,max_pool2d,clone,adaptive_avg_pool2d_out}`

Personally, I think the abbreviation `quant` is less clear than saying `quantized`. `quant` is a prefix that exists in many different words, and shortening it by 4 characters doesn't save much.

Additionally, I don't like trailing adjectives because that's not the word order in English. You don't say `equal quantized`, you say `quantized equal`. 

imo the convention should be quantized_xxxx

jerryzh168 replied:
it's mostly matching the existing conventions for cpu/cuda tensor type ids, xxx_cpu, and xxx_cuda

jamesr66a replied:
Additionally, trailing adjectives make the names inconsistent with the convention of {sparse,mkldnn,fbgemm,cudnn}_XXXX
